[1.1, 1.2단계 - 장기] 도기(이동엽) 미션 제출합니다.

dye0p opened the pull request:
안녕하세요 제이미! 이번 장기 미션의 리뷰이 도기 입니다! 
이번 미션의 리뷰어를 맡아 주셔서 감사합니다!
부족한 부분이 많지만 잘 부탁드립니다! 🙇‍♂️

<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
 - [셀프 체크](https://prolog.techcourse.co.kr/studylogs/4119)


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [x] 2
- [ ] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

- 각각의 기물의 이동 조건을 구현하는 과정에서 많은 분기문이 생겨났습니다. 
- 기물의 이동 조건에서 다른 기물들과 비슷한 모양의 중복 로직이 많이 발생하였습니다. 
- 컬렉션을 일급 컬렉션으로 포장하지 못했습니다. 포장할 일급 컬렉션이 많이 남아있습니다.
- 무분별하게 getter를 사용했습니다.
- 10줄이 넘는 메서드가 많습니다. 메서드 10줄 이하 원칙을 지키지 못했습니다.

  
## 어떤 부분에 집중하여 리뷰해야 할까요?

### 객체의 책임
객체들간의 책임이 적절하게 분배되고 있는지 의문이 들었습니다. 장기말의 책임은 자신이 목적지 까지 갈 수 있는지 여부를 판단하는 책임을 가지고 있습니다. 하지만 목적지를 가는 중에 다른 기물이나 장애물이 있는지 판단하는 책임은 Board에서 가지고 있습니다. 

- 각자의 말이 목적지 까지 이동가능한지 확인하는 책임 (장기 말)
- 목적지까지 가는데에 장애물이 있는지 판단하는 책임 (Board)

두가지의 책임을 각각의 장기말에서 수행하도록 하려 했지만 두번째 책임을 각각의 장기말이 가지게 되면 한칸을 이동하기 위해서 모든 장기말 클래스를 순회하면서 물어봐야 하는 일이 생겨버려 비효율 적이라 생각했습니다.

객체의 책임 판별에 대한 기준이 명확하게 서지 않는것 같습니다. 
제이미가 생각하는 객체의 책임 관점에서 리뷰 해주시면 좋을 것 같아요!

나머지 궁금한 사항은 코드 코멘트로 남겨놓겠습니다! 리뷰 잘 부탁 드립니다 : )

dye0p committed:
docs: 기능 명세서 작성

dye0p committed:
feat: 위치 클래스는 행과 열의 위치 정보를 가진다.

dye0p committed:
refactor: 네이밍 변경

dye0p committed:
feat: 기물은 이름과 위치를 가진다.

dye0p committed:
feat: 기물들의 위치와 이름을 세팅한다.

dye0p committed:
feat: 장기판의 범위를 초과하면 예외를 발생한다.

dye0p committed:
feat: 장기판은 기물과 위치를 가진다.

dye0p committed:
feat: 기물은 본인의 움직임을 계산할 수 있다.

dye0p committed:
feat: 기물의 이름과 초,한 개념을 가지는 프로필 객체 추가

dye0p committed:
refactor: 장기말들이 이동할 수 없는 위치라면 예외를 던지도록 변경

dye0p committed:
feat: 기물이 이동했다면 장기판의 기물들의 위치가 업데이트 된다.

dye0p committed:
feat: 기물은 자신이 가야하는 경로의 위치를 계산한다.

dye0p committed:
feat: 장기판은 가야할 위치에 장애물이 있다면 예외를 던진다.

dye0p committed:
feat: 장기 게임 입 출력 로직 구현

dye0p commented on src/main/java/chessPiece/Sang.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    @Override
+    public List<BoardPosition> makeRoute(final BoardPosition boardPosition) {
+        List<BoardPosition> route = new ArrayList<>();
```
`ChessPiece`의 추상 메서드인 `makeRoute`를 구현하는 각각의 기물은 자신의 위치에서 목적지 까지의 경로를 만들어서 반환합니다. 
상(Sang)같은 경우 갈 수 있는 경우의 수가 8가지어서 총 8번의 분기가 발생했습니다. 만약 갈 수 있는 경우의 수가 추가 된다면 더욱 많은 분기문이 생길것 같습니다. 

이를 해결하기 위한 방법으로 각각의 기물에 대한 경로를 `List로 반환하는 추상 메서드`를 가진 인터페이스를 구현하여 (예를들면 `RouteMakeStrategy` 인터페이스) 각각의 경우의 수를 계산하는 구현체를 만들어 사용하는 방법을 생각했습니다. 

이러한 방식의 단점은 각각의 이동 방향에 따른 객체가 생성되어 많은 클래스가 생성될 것 이라고 생각합니다.

❗️또 다른 생각이 떠올라 추가합니다!

각각의 이동 방향을 나타내는 Enum을 정의하고 이동 경로 계산 로직을 Enum에게 위임하는 방법입니다.
예를들어 UP_LEFT(3, 2)와 같은 상수를 정의하고 계산 로직을 부여할 수 있을 것 같습니다.

위와 같은 방법을 어떻게 생각하는지, 더 나은 설계를 위해서 어떤 방향으로 접근하면 좋을지 제이미의 생각이 궁금합니다!

jamie9504 replied:
두 방법 다 장/단점이 있으니 다 구현해보면 좋겠지만, 시간 관계상 생각했던 방식으로 설계한 크루가 있는지 찾아보고 의견을 나눠봐도 좋을 것 같네요.
[숙제] 각각의 방법으로 구현한 크루를 2명씩 찾아서, 의견을 나눠보고 정리 후 그 중 마음에 드는 방식으로 변경해보세요.

--- 

이와 별개로, 지금 움직이는 경로가 비슷한 Piece 친구들도 보이네요.
이 친구들은 어떻게 해줄 수 있을까요?

dye0p replied:
> 두 방법 다 장/단점이 있으니 다 구현해보면 좋겠지만, 시간 관계상 생각했던 방식으로 설계한 크루가 있는지 찾아보고 의견을 나눠봐도 좋을 것 같네요. [숙제] 각각의 방법으로 구현한 크루를 2명씩 찾아서, 의견을 나눠보고 정리 후 그 중 마음에 드는 방식으로 변경해보세요.

시간이 부족해 첫번째 방식(각각의 움직임을 인터페이스로 구현)으로 구현한 후 다른 크루들과 이야기를 해보았습니다.

장점
- 각각의 움직임을 구현체로 구현해 각각의 클래스에서 움직임을 정의할 수 있다.

단점
- 이동횟수에 따라서 클래스가 많이 생성된다.
- 하나의 기물에 대한 이동전략인데, 이것은 기물에 대한 이동전략을 추상화 하기보단 하나의 기물에 대한 이동 로직을 다른 객체에게 위임한 것으로 보인다.

결론적으로 장점보단 단점이 더 많은것 같은 느낌을 받았고, 

- 기물의 전체적인 이동을 추상화 하기 힘들다고 판단, 
- 각각의 기물에 대한 움직임을 인터페이스로 추상화 하는 것은 객체지향 적이라고 느끼지 못함(이동 전략이라는 추상화를 잘 못한 것으로 판단)

추상화 보단 책임을 다른 객체에게 위임한것으로 밖에 안보여 결국은 그대로 나두었습니다!

> 
> 이와 별개로, 지금 움직이는 경로가 비슷한 Piece 친구들도 보이네요. 이 친구들은 어떻게 해줄 수 있을까요?

현재 로직에서는 기물이 이동할 수 있는 각각의 방향에 대해서 명시적인 코드로 표현하고 있습니다. 이 부분에서 코드 중복을 줄이려면 수학적 계산이 조금 들어가는 부분이 있었습니다. 시간 관계상 이 부분을 제대로 생각하지 못하여 중복을 줄이기에 실패하였습니다.

하지만 지금 현재 코드의 장단점을 생각해보았을때

장점
- 방향에 대한 분기를 명시적으로 작성되어 있어, 한번에 알아보기 편함

단점
- 움직임이 비슷한 기물에 대한 중복로직이 발생
- 코드 라인수가 길어짐

이정도로 정리해보았습니다!

jamie9504 replied:
좋습니다.

> 이와 별개로, 지금 움직이는 경로가 비슷한 Piece 친구들도 보이네요. 이 친구들은 어떻게 해줄 수 있을까요?

이런 친구들은 조금 더 추상화를 해봐도 좋지 않을까요?
비슷한 경로를 움직이는 친구들끼리 묶을 수도 있을 것 같은데요.

dye0p commented on src/main/java/chessPiece/Board.java:
```diff
@@ -0,0 +1,137 @@
+
+public class Board {
+
+    private final Map<BoardPosition, ChessPiece> janggiPan = new HashMap<>();
```
현재 구조에서 Board는 Map을 통하여 구현하고 있습니다. Board를 구현하기 위해서 페어와 의논한 방법은 총 3가지 입니다.

### 1.  2차원 배열 사용

장점: 9 x 10 배열을 사용하여 각각의 기물을 인덱스로 관리해주려 했습니다. 배열을 꺼리는 이유는 크기가 가변적이지 못한 이유라고 생각합니다. 하지만 장기라는 도메인에서 장기판의 크기는 앞으로 바뀌지 않을 것이라는 생각을 했고 언급한 배열의 단점은 문제가 되지 않을것이라 생각했습니다. 

단점: String 배열이 아닌 ChessPiece 배열이기 때문에, 기물이 없는 칸에는 null이 들어갈 수 밖에 없습니다. 이 때문에 null값을 매번 검증해주어야 하는 단점이 있었습니다. 따라서 배열을 사용하는 것을 포기 하였습니다.

### 2. 리스트 사용
장점: 하나의 리스트에 초나라, 한나라의 기물을 모두 넣어 관리하고자 했습니다. 기물의 위치는 각각의 기물이 알고 있고, 리스트의 특징인 가변성을 활용하면 크기의 제약을 받지 않을 것 같다는 생각을 했습니다. 또한 null이 들어갈 수 있는 상황을 만들지 않아 null 에 대해서 안정성도 보장할 수 있습니다.

단점: 특정 기물을 찾기 위해서는 매번 모든 List를 순회해야 한다는 단점이 있었습니다. 장기 말의 이동이 잦은 도메인 특성상 매번 List를 순회하는 것은 성능상 전혀 이점이 없다 판단하였고, 따라서 리스트 사용을 포기 하였습니다.

### 3. Map 사용
장점: key = BoardPosition, value = CheesPiece 로 지정하여 키 벨류로 접근할 수 있습니다. key(BoardPosition) 으로 value(ChessPiece)를 찾기 때문에 매번 순회하던 리스트의 단점을 보완할 수 있고, null을 포함하지 않기에, null에 취약하던 배열의 단점을 보완할 수 있습니다.

단점: CheesPiece에서 BoardPosition을 상태로 가지고 있는 상태인데, Map에서도 BoardPosition을 가진다면 같은 상태를 중복해서 가진다고 생각이 들었습니다. 이는 타 개발자가 코드를 봤을때 Map의 구조를 보고 의문을 가질 수 있겠다는 생각이 들었습니다.
 
하지만 Map을 사용했을때 배열과 리스트의 단점을 보완한다는 점, key value 형태로 값을 편하게 꺼낼 수 있다는 장점이 크게 느껴졌고, 정답은 아니지만 가장 최선의 방법이라는 합리적인 판단에 Map을 사용하기로 결정했습니다.

위와같은 결정에 대한 제이미의 생각은 어떤지 궁금합니다. 또한 이러한 고민과 결정에 대해서 납득할만한 근거가 충분한지 궁금합니다!

jamie9504 replied:
충분히 잘 정리했다고 생각하는데, 근거가 불충분했다고 느끼셨나요?

도기는 저를 납득시키고 싶으신건가요?
아니면 아래 중 어떤 사람을 납득시키고 싶은건가요?
각 사람마다 난이도와 근거의 준비량이 달라지겠죠?

1. 도기와 비슷한 사고방식을 가진 누군가
2. 도기와 다른 사고방식을 가진 누군가
3. 도기의 방식을 이해하지 못한 누군가
4. 도기의 방식이던 다른 방식이던 상관 없는 누군가

---

별개로 적은 내용을 조금 더 정제해보는 건 어떨까요?
지금은 쭉 줄 글로 나열한 느낌인데, 보고서 느낌으로 명료하게 정리해보세요.

dye0p replied:
> 도기는 저를 납득시키고 싶으신건가요?

표현을 제가 애매하게 한 것 같습니다! 납득보단 제가 작성한 이유와 근거가 제이미에게 충분히 잘 설명이 된 것 같은지 궁금했습니다!

> 별개로 적은 내용을 조금 더 정제해보는 건 어떨까요?
지금은 쭉 줄 글로 나열한 느낌인데, 보고서 느낌으로 명료하게 정리해보세요.

리팩터링을 진행하면서 결국은 Map 컬렉션을 사용하는 방식을 선택했습니다.

장점

- 기물이 가지고 있던 포지션 상태를 제거하고 Map에서 key = Position, value = Piece 으로 한번에 관리가 가능하다.
-  Map의 특성으로 위치에 따른 기물의 접근과 수정이 편리하다.

단점 (정확히 단점인지는 잘 모르겠습니다. 개개인마다 견해가 다를거라고 생각합니다!)
- 기물에서 상태였던 Position을 제거하였는데, 기물이 Position을 가지고 있다면 각각의 기물이 자신의 위치를 관리하여 조금 더 능동적인 객체가 될 수 있을 것 같은데, Map을 사용함으로써 기물이 능동적인 객체인지 의문이 들었음.

결론적으로 Map으로 관리하는것이 코드의 복잡성, 관리의 용이 측면에서 더 이득이 클 것이라고 판단하였습니다!

jamie9504 replied:
그런 관점이라면 충분히 잘 설명이 되었다고 생각해요.
개인적으론 줄글보다는 명료한 글을 선호하는데, 이건 사람마다 취향 차이가 있으니 둘 다 쓸 수 있으면 좋겠죠!

dye0p reviewed (commented):
안녕하세요 제이미! 코드에 관한 고민 사항을 남겨보았습니다! 리뷰 잘 부탁드립니다 :)

jamie9504 commented on README.md:
```diff
@@ -1,3 +1,59 @@
 # java-janggi
 
 장기 미션 저장소
+
```
> 각각의 기물의 이동 조건을 구현하는 과정에서 많은 분기문이 생겨났습니다.
> 기물의 이동 조건에서 다른 기물들과 비슷한 모양의 중복 로직이 많이 발생하였습니다.
> 컬렉션을 일급 컬렉션으로 포장하지 못했습니다. 포장할 일급 컬렉션이 많이 남아있습니다.
> 무분별하게 getter를 사용했습니다.
> 10줄이 넘는 메서드가 많습니다. 메서드 10줄 이하 원칙을 지키지 못했습니다.

개선해볼까요?
프로그래밍 요구 사항 모두 챙겨주세요.

> 객체의 책임 판별에 대한 기준이 명확하게 서지 않는것 같습니다.

`객체지향의 사실과 오해`를 읽으셨나요?
읽으셨다면 한 번 더 훑어보고, 읽으시지 않았다면 한 번 읽으신 후 느낀 점을 정리해주세요.

> 안녕하세요 제이미! 코드에 관한 고민 사항을 남겨보았습니다! 리뷰 잘 부탁드립니다 :)

이 부분 리뷰 요청 이후 작성된 코멘트로 보이네요.
다음 리뷰 때 확인하겠습니다.

리뷰 요청 후는 PR 본문, 코멘트, 코드 모두 추가 수정이 되지 않아야 합니다.

jamie9504 commented on src/main/java/chessPiece/Board.java:
```diff
@@ -0,0 +1,137 @@
+        int cnt = 0;
+
+        if (dx == 0 && dy > 0) {
+            for (int i = 1; i <= dy; i++) {
```
단위 테스트가 되지 않은 부분들이 보이네요.

---

`!` not으로만 쓸거면, not의 의미를 메서드명이 담는게 더 가독성이 좋겠죠?

dye0p replied:
> 단위 테스트가 되지 않은 부분들이 보이네요.

모든 분기를 테스트하지 못한 것 같습니다. 테스트 추가하였습니다!

jamie9504 commented on src/main/java/Application.java:
```diff
@@ -0,0 +1,28 @@
+import view.OutputView;
+
+public class Application {
+    public static void main(String[] args) {
```
해당 에러메시지는 적합한 에러메시지인가요?
예외가 발생했을 때, 사용자에게 StackTrace를 노출하는 이유는 무엇인가요?

```
ㅤㅤㅤ 0ㅤㅤ 1ㅤㅤ 2ㅤㅤ 3ㅤㅤ 4ㅤㅤ 5ㅤㅤ 6ㅤㅤ 7ㅤㅤ 8ㅤ
0ㅤ | 차 | 상 | 마 | 사 | ㅤ | 사 | 마 | 상 | 차
1ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | 왕 | ㅤ | ㅤ | ㅤ | ㅤ
2ㅤ | ㅤ | 포 | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | 포 | ㅤ
3ㅤ | 병 | ㅤ | 병 | ㅤ | 병 | ㅤ | 병 | ㅤ | 병
4ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ
5ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ
6ㅤ | 졸 | ㅤ | 졸 | ㅤ | 졸 | ㅤ | 졸 | ㅤ | 졸
7ㅤ | ㅤ | 포 | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | 포 | ㅤ
8ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | 왕 | ㅤ | ㅤ | ㅤ | ㅤ
9ㅤ | 차 | 상 | 마 | 사 | ㅤ | 사 | 마 | 상 | 차
초나라 턴 입니다. 움직이고 싶은 말의 위치를 입력해주세요.
09
Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: Index 1 out of bounds for length 1
```

---

오타 체크해주세요.

장기를 둘 수 있는 사용자가 해당 콘솔을 실행시키면
말의 위치를 어떻게 입력하는지 알 수 있나요?

사용자에게 친절한 프로그램인가요?

```
초나라 턴 입니다. 움직이고 싶은 말의 위치를 입력해주세요.
9,0
해당 말을 이동시킬 위치를 입려해 주세요.
9,8
[ERROR] 이동하려는 경로에 장애물이 존재합니다.
```

---

적합한 에러메시지인가요?
전반적으로 검토해주세요.

```
ㅤㅤㅤ 0ㅤㅤ 1ㅤㅤ 2ㅤㅤ 3ㅤㅤ 4ㅤㅤ 5ㅤㅤ 6ㅤㅤ 7ㅤㅤ 8ㅤ
0ㅤ | 차 | 상 | 마 | 사 | ㅤ | 사 | 마 | 상 | 차
1ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | 왕 | ㅤ | ㅤ | ㅤ | ㅤ
2ㅤ | ㅤ | 포 | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | 포 | ㅤ
3ㅤ | 병 | ㅤ | 병 | ㅤ | 병 | ㅤ | 병 | ㅤ | 병
4ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ
5ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ
6ㅤ | 졸 | ㅤ | 졸 | ㅤ | 졸 | ㅤ | 졸 | ㅤ | 졸
7ㅤ | ㅤ | 포 | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | 포 | ㅤ
8ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | 왕 | ㅤ | ㅤ | ㅤ | ㅤ
9ㅤ | 차 | 상 | 마 | 사 | ㅤ | 사 | 마 | 상 | 차
초나라 턴 입니다. 움직이고 싶은 말의 위치를 입력해주세요.
0,0
해당 말을 이동시킬 위치를 입려해 주세요.
0,1
[ERROR] 이동하려는 경로에 장애물이 존재합니다.

---

11, 13
[ERROR] 좌표는 숫자만 입력 할 수 있습니다.
```

dye0p replied:
> 해당 에러메시지는 적합한 에러메시지인가요?
예외가 발생했을 때, 사용자에게 StackTrace를 노출하는 이유는 무엇인가요?

구현 하는 과정에서 확인을 꼼꼼히 확인하지 못한 것 같습니다..! 수정하겠습니다!

> 오타 체크해주세요.

오타 확인 했습니다!

> 장기를 둘 수 있는 사용자가 해당 콘솔을 실행시키면
> 말의 위치를 어떻게 입력하는지 알 수 있나요?

> 사용자에게 친절한 프로그램인가요?

클라이언트의 입장에서 볼때 어떻게 장기말을 입력해야 하는지 판단하기 어려울 것이라고 생각이 드네요! 클라이언트의 입장에서 이해 하기 쉽도록 수정하겠습니다!

> 적합한 에러메시지인가요?
> 전반적으로 검토해주세요.

수정하겠습니다!

jamie9504 commented on src/main/java/Application.java:
```diff
@@ -0,0 +1,28 @@
+    public static void main(String[] args) {
+
+        ChessPieceInitializer chessPieceInitializer = new ChessPieceInitializer();
+        Board board = new Board(chessPieceInitializer.hanInit(), chessPieceInitializer.choInit());
```
메서드명으로 적합한 메서드명인가요?
전반적인 검토해주세요.

dye0p replied:
직관적으로 이해하지 못하는것 같아요! 수정했습니다!

jamie9504 commented on src/main/java/chessPiece/Board.java:
```diff
@@ -0,0 +1,137 @@
+package chessPiece;
```
패키지를 분리한 기준이 무엇인가요?

dye0p replied:
장기말들의 클래스가 많아짐에 따라서 하나의 chesspiece의 개념으로 묶고자 했습니다! 지금 보니 패키지명은 소문자로 작성되어야 하는데 카멜 케이스로 작성 되었네요! 패키지 네이밍도 장기 도메인과 어울리지 않는것 같습니다. 관련된 클래스들끼리 모이도록 패키지를 구분했습니다!

jamie9504 commented on src/main/java/chessPiece/Board.java:
```diff
@@ -0,0 +1,137 @@
+
+public class Board {
+
+    private final Map<BoardPosition, ChessPiece> janggiPan = new HashMap<>();
```
chessPiece, janggiPan 같은 기준으로 분류된 용어인가요?

dye0p replied:
코드를 처음 보는 개발자의 입장에서 헷갈릴것 같네요! 도메인에 맞는 용어로 수정했습니다!

[refactor: 기물 추상 클래스 네이밍 변경(chessPiece -> Piece)](https://github.com/woowacourse/java-janggi/pull/23/commits/aa93fb2224b334bdc1c67eadfb68a08471ab09df)
[refactor: 장기판 관련 변수 네이밍 변경(janggipan -> janggiBoard)](https://github.com/woowacourse/java-janggi/pull/23/commits/840ade192f2b75c53ff3477890025898161128bf)

jamie9504 commented on src/main/java/chessPiece/Board.java:
```diff
@@ -0,0 +1,137 @@
+        );
+    }
+
+    public Map<BoardPosition, ChessPiece> getJanggiPan() {
```
도기는 클래스 내 메서드 선언 순서가 어떻게 되나요?

dye0p replied:
저는 보통 public -> private 순으로 배치하는 것 같습니다. 세부적으로 들어가면

### public 메서드
- private 메서드 
- private 메서드

처럼 하나의 public 메서드를 상단에 선언하고, 해당 public 메서드안에서 호출되는 private메서드를 호출되는 순서로 바로 아래에 배치합니다.

public 메서드 끼리의 순서는

`상태변경(1) - 판별(2,3) -> 조회(2,3)`
순으로 배치하는 편입니다.

그리고 `getter()/setter()`, `toString()`, `equals() & hashCode()` 등과 같은 메서드는 가장 아래 하단에 배치하는 편 입니다!

급하게 코드를 작성하다보니 메서드의 배치순서가 엉망이 되버린것 같습니다. 컨벤션을 지켜 수정했습니다!

jamie9504 commented on src/main/java/chessPiece/Byeong.java:
```diff
@@ -0,0 +1,32 @@
+
+import java.util.List;
+
+public class Byeong extends ChessPiece {
```
장기 기획서에 영어로 표기된 말들이 있는데, 한문명을 그대로 사용한 이유는 무엇인가요?

dye0p replied:
한문명을 그대로 사용하면 클래스명을 보고 직관적으로 읽을 수 있을 것이라 생각했습니다!  이런 형태의 클래스 네이밍에 대한 제이미의 생각도 궁금합니다!

jamie9504 replied:
한문명을 쓰던, 영문명을 쓰던 일관성이 있으면 직관적일 수 있죠.

Chess Piece는 한문명인가요?
클래스명 명명 방식이 일관성이 있었나요?

dye0p replied:
맞습니다! 이전의 네이밍은 일관성이 없었다는 문제가 있었습니다. 기존의 한문 네이밍에서 영문 네이밍으로 수정하였습니다! 

[refactor: 기물 네이밍 영문 네이밍으로 변경](https://github.com/woowacourse/java-janggi/pull/23/commits/913d0837753d6728bc1f7141a342b6807e0aaeb7)

jamie9504 commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,39 @@
+    private final Scanner scanner = new Scanner(System.in);
+
+    public BoardPosition readPresentPick() {
+        System.out.println("초나라 턴 입니다. 움직이고 싶은 말의 위치를 입력해주세요.");
```
해당 문제들이 발생한 이유는 무엇인가요?

미션 리뷰 요청시 프로그램은 어떤 상태여야 할지 도기의 의견을 코멘트로 정리해주세요.

```
ㅤㅤㅤ 0ㅤㅤ 1ㅤㅤ 2ㅤㅤ 3ㅤㅤ 4ㅤㅤ 5ㅤㅤ 6ㅤㅤ 7ㅤㅤ 8ㅤ
0ㅤ | 차 | 상 | 마 | 사 | ㅤ | 사 | 마 | 상 | 차
1ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | 왕 | ㅤ | ㅤ | ㅤ | ㅤ
2ㅤ | ㅤ | 포 | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | 포 | ㅤ
3ㅤ | 병 | ㅤ | 병 | ㅤ | 병 | ㅤ | 병 | ㅤ | 병
4ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ
5ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ
6ㅤ | 졸 | ㅤ | 졸 | ㅤ | 졸 | ㅤ | 졸 | ㅤ | 졸
7ㅤ | ㅤ | 포 | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | 포 | ㅤ
8ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | 왕 | ㅤ | ㅤ | ㅤ | ㅤ
9ㅤ | 차 | 상 | 마 | 사 | ㅤ | 사 | 마 | 상 | 차
초나라 턴 입니다. 움직이고 싶은 말의 위치를 입력해주세요.
6,0
해당 말을 이동시킬 위치를 입려해 주세요.
5,0
ㅤㅤㅤ 0ㅤㅤ 1ㅤㅤ 2ㅤㅤ 3ㅤㅤ 4ㅤㅤ 5ㅤㅤ 6ㅤㅤ 7ㅤㅤ 8ㅤ
0ㅤ | 차 | 상 | 마 | 사 | ㅤ | 사 | 마 | 상 | 차
1ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | 왕 | ㅤ | ㅤ | ㅤ | ㅤ
2ㅤ | ㅤ | 포 | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | 포 | ㅤ
3ㅤ | 병 | ㅤ | 병 | ㅤ | 병 | ㅤ | 병 | ㅤ | 병
4ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ
5ㅤ | 졸 | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ
6ㅤ | 졸 | ㅤ | 졸 | ㅤ | 졸 | ㅤ | 졸 | ㅤ | 졸
7ㅤ | ㅤ | 포 | ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | 포 | ㅤ
8ㅤ | ㅤ | ㅤ | ㅤ | ㅤ | 왕 | ㅤ | ㅤ | ㅤ | ㅤ
9ㅤ | 차 | 상 | 마 | 사 | ㅤ | 사 | 마 | 상 | 차
초나라 턴 입니다. 움직이고 싶은 말의 위치를 입력해주세요.
```

dye0p replied:
리뷰 요청시 프로그램의 상태는
1. 모든 테스트가 정상 동작하는 상태
2. 제시된 요구사항을 준수하는 상태
3. 프로그램이 정상 실행되는 상태

여야 한다고 생각합니다.

그러기 위해선 pr을 보내기 전, 프로그램에 오류가 없는지 확인 하는 것이 중요하다고 생각합니다. 

이번 리뷰 요청에는 제대로 확인하지 못한 채 제출한 것 같습니다. 수정완료 하였습니다!

jamie9504 reviewed (changes requested):
안녕하세요. 도기, 리뷰어 제이미입니다.

만나서 반갑습니다.

코멘트 남겨두었습니다.

dye0p committed:
refactor: 입력 뷰 리팩터링

- 메서드 분리
- 매직넘버 & 매직 스트링 상수화
- 입력 좌표 갯수 검증 추가

dye0p committed:
fix: 장기말 이동 후, 이동 전 말들이 사라지지 않는 문제 해결

dye0p committed:
refactor: 장기말 입력 메세지 직관적으로 수정

dye0p committed:
refactor: 패키지 세분화 후 이동

dye0p committed:
feat: 장기게임 실행 클래스 추가

- 장기 게임을 실행한다
- 잘못된 입력을 하면 예외를 던지고 재입력 한다.

dye0p committed:
refactor: 기물 추상 클래스 네이밍 변경(chessPiece -> Piece)

- 장기 도메인 개념과 관련된 용어로 변경

dye0p committed:
refactor: 피스 추상 클래스 생성자 접근 제어자 변경(public -> protected)

- 생성자가 public으로 열려있다면 외부에서 생성 가능하다는 오해를 줄 수 있다.
- 하위 타입에서만 생성 가능하다는 의도를 명확히 하기 위해서 생성자 접근제어자를 변경한다.

dye0p committed:
refactor: 장기판 관련 변수 네이밍 변경(janggipan -> janggiBoard)

dye0p committed:
refactor: 포지션 클래스 네이밍 변경(BoardPosition -> Position)

dye0p committed:
refactor: 장기판 getter 메서드 불변으로 반환하도록 변경

dye0p committed:
refactor: 장기판 기물 초기화 방식 변경

- 기존의 han, cho 두가지의 리스트를 각각 생성해 장기판을 생성하던 방식에서 하나의 리스트에 모든 기물을 담아 장기판을 생성하는 방식으로 변경

dye0p committed:
feat: 포 이동 조건 추가

- 포는 반드시 포를 제외한 기물 하나를 넘어가야 한다.

dye0p committed:
refactor: 메서드 네이밍을 명확하게 변경

dye0p committed:
refactor: 장기판 로직 메서드 정리

- 중복 메서드 분리
- 사용하지 않는 메서드 제거

dye0p committed:
refactor: 장애물 판단 로직 각각의 기물로 책임 이동

dye0p committed:
feat: 같은 팀의 기물을 잡을 수 없는 기능 구현

dye0p committed:
test: 차 이동 테스트 추가

dye0p committed:
test: 마 이동 테스트 추가

dye0p committed:
fix: 사 이동 예외 메세지 수정

dye0p committed:
test: 상 이동 테스트 추가

dye0p committed:
test: 병 이동 테스트 추가

dye0p committed:
refactor: 변경될 가능성이 없는 매개변수, 지역변수 final 키워드 추가

dye0p committed:
test: 상 방향별 이동 경로 계산 테스트 추가

dye0p committed:
test: 차 방향별 이동 경로 계산 테스트 추가

dye0p committed:
test: 졸 이동 경로 장애물 테스트 추가

dye0p committed:
test: 마 방향별 이동경로 계산 테스트 추가

dye0p committed:
test: 포 방향별 이동경로 계산 테스트 추가

dye0p committed:
test: 사 이동 경로 장애물 테스트 추가

dye0p committed:
test: 테스트 코드 변수 final 키워드 추가

dye0p committed:
refactor: 포지션 업데이트 로직 추상 클래스로 이동

- 모든 기물에서 동일하게 포지션 업데이트 로직을 사용하므로 하나의 추상으로 이동시켜 코드 중복을 제거함

dye0p committed:
feat: 차례 턴 기능 추가

- 한 차례식 턴을 바꾸며 게임을 할 수 있다.

dye0p committed:
feat: 선택한 위치에 기물이 존재하지 않는다면 예외가 발생한다.

dye0p committed:
refactor: 기물 이동 성공 메세지 출력 개행 추가

dye0p committed:
refactor: 장기판 크기 매직넘버 상수화

dye0p committed:
test: 다른 팀의 기물이라면 예외발생 테스트 추가

dye0p committed:
refactor: 이동 경로 계산 로직 방향별 메서드 분리

dye0p committed:
refactor: 기물 이동 과정 piece 객체로 이동

- 시간적 결합 해소

dye0p committed:
feat: 상대편 기물을 잡을 수 있다.

- 목적지 직전까지의 경로를 탐색하여 장애물 없이 목적지 직전까지 왔다면 목적지에 상대편 기물이 있다면 잡는다.

dye0p reviewed (commented):
안녕하세요 제이미! 처음 구현과정에서 미흡한 부분이 많아, 추가할 부분이 많아서 피드백 반영이 늦어졌습니다! 리뷰 잘 부탁드립니다 : )

jamie9504 commented on src/main/java/janggi/piece/Byeong.java:
```diff
@@ -0,0 +1,43 @@
+        }
+    }
+
+    private void validateObstacle(final Map<Position, Piece> janggiBoard, final Position position) {
```
테스트가 되고 있나요?

dye0p replied:
기물들 중 한칸만 이동할 수 있는 기물들이 있었습니다 (병, 왕, 졸, 사)
한칸만 이동하기에 가고자 하는 경로에 장애물을 검증할 필요가 없었습니다. 
하지만 메서드가 존재했던 이유는 Piece 추상 클래스에서 장애물 검증 메서드를 추상 메서드로 정의하고 있었기 때문입니다. 

이를 해결하기 위해서 기존의 추상 메서드였던 `checkObstacle` 메서드를 변경해주었습니다.

``` java
   protected abstract void checkObstacle(final Position targetPosition, final Map<Position, Piece> janggiBoard);
```

``` java
   protected void checkObstacle(final Position targetPosition, final Map<Position, Piece> janggiBoard) {
          // 빈 껍데기를 가진 메서드
    }
```
추상 클래스에서 빈 껍데기를 가진 디폴트 메서드를 선언하고, 장애물 검증이 필요한 구현체에만 구현하여 사용할 수 있도록 하였습니다. 이렇게 함으로써 필요하지 않은 메서드를 지울 수 있었고, 해당 메서드는 테스트 할 필요가 없어, 테스트를 지울 수 있었습니다.

이러한 방식이 정답은 아니라고 생각이 들어 제가 생각하는 장단점을 정리해보았습니다.

장점 

- 추상 메서드로 구현하지 않아, 해당 메서드가 필요한 구현체에서만 사용할 수 있다.
- 위의 이유로 필요하지 않은 구현체는 강제로 구현하지 않아도 된다.
- 결론적으로 코드의 유연성과 확장성이 증가한다고 생각한다,

단점

- 추상 클래스의 메서드는 모든 자식에서 동일하게 사용할 수 있어야 한다고 생각한다.
   -  이 부분에 대해서는 '추상 메서드에 한정해서만 모든 구현체에서 동일하게 사용이 가능해야 한건지?' 와 헷갈리는 부분이 있는것 같습니다.

- 선언된 메서드가 빈 껍데기이기 때문에 의도를 파악하기 어려움
   - 빈 메서드로 존재하기 때문에 "이 메서드는 무엇을 하는 것인지" 헷갈릴 수 있다고 생각함
- 모든 자식 구현체에서 사용하지 않아도 되기 때문에 정확히 '어떤 서브 클래스'에서 오버라이드 할지 명확하지 않음

이정도로 정리해 볼 수 있을 것 같습니다! 혹시 부족한 부분이 있거나 추가적으로 학습해야할 부분이 있는지 궁금합니다!

[refactor: 한칸만 이동하는 기물과 여러칸을 이동하는 기물을 별도의 패키지로 구분](https://github.com/woowacourse/java-janggi/pull/23/commits/4069b528416f47ad4ab01417ca1b04d850358a2c)
[refactor: 한칸만 이동하는 기물의 장애물 검사 로직 제거](https://github.com/woowacourse/java-janggi/pull/23/commits/5cb9a9bfab0b9593a87e505e1578cdbb6144e71d)

jamie9504 replied:
추상화라는 개념이, 도기가 남겨준 단점을 안고 가도 된다고 생각하는지에 대한 의견이 궁금하네요.
추상화된 메서드를 다 사용하지 않는다면, 잘 추상화가 된 걸까요?

---

인터페이스 분리 원칙(ISP), 전략 패턴(Strategy Pattern), 템플릿 메서드 패턴(Template Method Pattern)에 대해서도 함께 학습하기 좋아보이는 부분이네요.

jamie9504 commented on src/main/java/janggi/JanggiGame.java:
```diff
@@ -0,0 +1,87 @@
+
+    private void playJanggi(final Board board) {
+        Team currentTurnTeam = FIRST_TURN_TEAM;
+        while (true) {
+            playTurn(board, currentTurnTeam);
+            currentTurnTeam = changeTurn(currentTurnTeam);
+        }
```
게임을 끝낼 순 없나요?

게임은 영원히 계속되나요?

dye0p replied:
게임 종료를 고려하지 못했네요! 왕이 잡히면 게임이 종료되도록 하였습니다!😀

[feat: 왕이 죽으면 게임이 종료된다.](https://github.com/woowacourse/java-janggi/pull/23/commits/fd967ddf51e692d693eb42a183829cd646039cc0)

jamie9504 commented on src/main/java/janggi/board/Board.java:
```diff
@@ -0,0 +1,47 @@
+    private final Map<Position, Piece> janggiBoard = new HashMap<>();
+
+    public Board(final List<Piece> pieces) {
+        janggiBoard.putAll(
+                pieces.stream()
+                        .collect((Collectors.toMap(Piece::getBoardPosition, piece -> piece))
+                        ));
```
필드 선언시 `new HashMap<>();`으로 생성하고, 생성자에서는 `putAll`을 하는 이유는 무엇인가요?

dye0p replied:
원래 로직에서 리스트로 기물을 생성하고 있었기 때문에 저렇게 짜여졌습니다. 조금 더 생각해봤을 때 결론적으로 Map으로 사용되는데 저러한 방식은 불필요한 방식이라고 생각했고, 

외부에서 Position 과 Piece를 생성하여 Board가 자신의 필드에 하나씩 넣어주도록 하였습니다. 

외부에서 Map을 생성하여 주입하는 방식보단, Board가 직접 Map에 넣으면 능동적일 거라고 생각을 하여 Board가 직접 넣어주도록 하였습니다! 

[refactor: 기물 초기화 방식 변경](https://github.com/woowacourse/java-janggi/pull/23/commits/354c2240ab3acf0582f7b2c4e10371124706c76a)

jamie9504 commented on src/main/java/janggi/board/PieceInitializer.java:
```diff
@@ -0,0 +1,72 @@
+    public List<Piece> generate() {
+        final List<Piece> pieces = new ArrayList<>();
+
+        pieces.add(new Cha(new PieceProfile("차", Team.HAN), new Position(0, 0)));
```
PieceProfile의 name을 꼭 항상 입력받아야 할까요?

dye0p replied:
이름을 외부에서 직접 정의해주게되면 변경에 취약하다는 문제가 있을 거라고 생각합니다. 외부에서 기물의 이름을 변경하면 막을 수 없기에 기물 내부에서 자신의 이름을 가진 타입을 생성자에서 정의해주도록 변경하였습니다!

[refactor: 기물 생성시 자신의 타입을 내부에서 지정하여 생성하도록 변경](https://github.com/woowacourse/java-janggi/pull/23/commits/c96efcbabdfa5c1c1d360b103d1b3b18bc459885)

jamie9504 commented on src/main/java/janggi/position/Position.java:
```diff
@@ -0,0 +1,46 @@
+    private final int col;
+
+    public Position(final int row, final int col) {
+        validateOutOfBound(row, col);
```
음수가 들어온다면 괜찮나요?

dye0p replied:
음수 입력에 대한 검증이 빠졌었네요! 추가하였습니다!

[feat: 위치 값이 음수인 경우 예외처리 추가](https://github.com/woowacourse/java-janggi/pull/23/commits/7f158a459417709f5de0b81d4f254b3cfc402ec4)

jamie9504 commented on src/test/java/janggi/piece/ChaTest.java:
```diff
@@ -0,0 +1,168 @@
+
+class ChaTest {
+
+    @DisplayName("차은 위치 정보를 가진다,")
```
```suggestion
    @DisplayName("차는 위치 정보를 가진다.")
```

Test DisplayName도 일관성이 있으면 좋습니다.

기물명으로 시작하는 경우도 있고 아닌 경우도 있네요.

jamie9504 reviewed (changes requested):
안녕하세요, 도기. 리뷰어 제이미입니다.

코멘트 남겨두었습니다.

다음 요청시 Approve & Merge 후 2단계 진행할 예정입니다.

dye0p committed:
feat: 위치 값이 음수인 경우 예외처리 추가

dye0p committed:
refactor: 기물 생성시 자신의 타입을 내부에서 지정하여 생성하도록 변경

- 기물의 이름을 외부에서 직접 지정해주었을떄 외부에서 이름이 변경될 가능성이 있었음
- 기물의 내부에서 자신의 기물 타입을 지정하여 생성하도록 변경하여 외부에서의 변경 가능성을 제거함

dye0p committed:
feat: 왕이 죽으면 게임이 종료된다.

dye0p committed:
refactor: 한칸만 이동하는 기물의 장애물 검사 로직 제거

- 추상 클래스 Piece 에서 추상 메서드로 선언되어 있던 장애물 검사 메서드를 빈 껍데기를 가진 일반 메서드로 변경하고, 장애물 검사가 필요한 구현체에서 오버라이드 하여 사용하도록 변경

dye0p committed:
refactor: 한칸만 이동하는 기물과 여러칸을 이동하는 기물을 별도의 패키지로 구분

dye0p committed:
refactor: 기물 네이밍 영문 네이밍으로 변경

dye0p committed:
refactor: Team 초나라 상수명 변경(CHO -> CHU)

dye0p committed:
refactor: 이동 가능 여부 메서드 좌표 계산 로직 Position 객체로 책임 이동

dye0p committed:
test: 포지션 테스트 추가

dye0p committed:
refactor: 기물 초기화 방식 변경

dye0p committed:
refactor: 기물의 상태 Position 필드 제거

dye0p committed:
refactor: 보드 초기화 객체 네이밍 변경(PieceInitializer -> BoardGenerator)

dye0p committed:
refactor: 위치 관련 변수 네이밍 변경

- presentPosition ->currentPosition

- futurePosition -> targetPosition

dye0p committed:
refactor: PieceProfile 객체 제거

dye0p reviewed (commented):
안녕하세요 제이미! 변경사항보단 고민에 대한 시간이 오래결려 피드백 반영이 늦어졌습니다! 남겨주신 피드백과 저의 생각을 정리해보았습니다! 리뷰 잘 부탁드립니다 :) 🙇‍♂️

jamie9504 commented on README.md:
```diff
@@ -1,3 +1,72 @@
 장기 미션 저장소
+
+## 입력
+
```
// todo

jamie9504 commented on src/main/java/janggi/piece/Piece.java:
```diff
@@ -0,0 +1,86 @@
+        this.team = team;
+    }
+
+    public void moveTo(final Position currentPosition, final Position targetPosition,
```
moveTo라는 이름만 봤을 땐 무언가 움직일 것 같은 느낌이 들죠?

jamie9504 commented on src/main/java/janggi/piece/Team.java:
```diff
@@ -0,0 +1,41 @@
+
+    public abstract Team changeTeam();
+
+    public static boolean isChu(final Team team) {
+        return CHU.equals(team);
+    }
+
+    public static boolean isHan(final Team team) {
+        return HAN.equals(team);
+    }
+
+    public boolean isSameTeam(final Team team) {
+        return this.equals(team);
+    }
```
밖에서 직접 비교하는 것과 어떤 차이가 있나요?

---

Enum 비교시 equals와 `==` 비교는 어떤 차이가 있나요?

jamie9504 commented on src/main/java/janggi/piece/multiplemovepiece/Cannon.java:
```diff
@@ -0,0 +1,129 @@
+    private void validatePoMove(final List<Position> moveRoute, final Map<Position, Piece> janggiBoard) {
+        int obstacleCount = 0;
+
+        for (final Position position : moveRoute) {
```
> 1️⃣ 한 메서드에 오직 한 단계의 들여쓰기만 한다.

jamie9504 commented on src/main/java/janggi/view/OutputView.java:
```diff
@@ -0,0 +1,92 @@
+    private static final String GREEN_COLOR_CODE = "\u001B[32m";
+    private static final String EXIT_CODE = "\u001B[0m";
+
+    private static final String EMPTY_SPACE = "ㅤ";
```
매직넘버를 상수화하는 이유는 무엇인가요

jamie9504 reviewed (approved):
안녕하세요, 도기. 리뷰어 제이미입니다.

코멘트 남겨두었습니다.

1단계 요구사항 충족되어 Approve & Merge 합니다.
남긴 코멘트에 대해 추가적으로 이야기할 부분이 있다면 2단계 PR 본문에 작성해주시기 바랍니다.
머지된 PR은 다시 확인하지 않습니다.
